File: lib_reducer/python/libreducer.py
#!/usr/bin/python26
import sys
from abc import ABCMeta,abstractmethod

class BaseReducer:
	__metaclass__ = ABCMeta
	
	def setup(self):
		pass
		
	def cleanup(self):
		pass
		
	def parse(self, line):
		kv = line.split("\t", 1)
		key = kv[0]
		value = "" if len(kv) == 1 else kv[1]
		return (key, value)
	
	@abstractmethod	
	def reduce(self, key, values):
		pass

class ValuesIterator:
	def __init__(self, reducer):
		self.current_key = "\t"
		self.last_key = "\t"
		self.is_eof = False
		self.is_newline = False
		self.init = False
		self.reducer = reducer
		
	def next_key(self):
		if not self.is_newline:
			while True:
				try: self.next()
				except StopIteration:
					if self.is_eof:
						return False
					else:
						self.is_newline = False
						return True
				if self.last_key != self.current_key:
					self.is_newline = False
					return True
		else: return True
			
	
	def __iter__(self):
		self.last_key = self.current_key
		self.is_newline = False
		self.init = True
		return self
		
	def next(self):
		if self.init:
			self.init = False
			return self.current_value
		while True:
			line = sys.stdin.readline()
			if line == "":
				self.is_eof = True
				raise StopIteration
			line = line.strip()
			if line != "": break		
		self.last_key = self.current_key
		(self.current_key, self.current_value) = self.reducer.parse(line)
		if self.last_key != "\t" and self.last_key != self.current_key:	
			self.is_newline = True
			raise StopIteration
		return self.current_value
		
		
def run(reducer):	
	reducer.setup()	
	values = ValuesIterator(reducer)
	while True:
		if not values.next_key(): break
		reducer.reduce(values.current_key, values)
	reducer.cleanup()
	

File: lib_reducer/php/libreducer.php
<?php
class ValuesIterator implements Iterator{
	private $last_key;
	private $current_key;
	private $current_value;
	private $is_newline;
	private $is_eof;
	
	public function __construct($reducer){
		$this->last_key = "\t";
		$this->current_key = "\t";
		$this->is_newline = false;
		$this->is_eof = false;
		$this->reducer = $reducer;
	}
	public function next_key(){
		if(!$this->is_newline){	
			do{
				$this->next();
			}while(!$this->is_eof && $this->last_key === $this->current_key);
			$this->is_newline = false;
		}
		return !$this->is_eof;
	}
	
	
	
	public function rewind() { 
		$this->last_key = $this->current_key;
		$this->is_newline = false;
	}

    public function current() {  
		return $this->current_value;
	}

    public function key() { 
		return $this->current_key;
	}

    public function next() {  
		do{
			$in = fgets(STDIN);
			if($in === false) {
				$this->is_eof = true;
				return;
			}
			$in = trim($in);
		}while($in === "");
		list($key, $value) = $this->reducer->parse($in);
		$this->last_key = $this->current_key;
		$this->current_key = $key;
		$this->current_value = $value;
		//echo $this->last_key . ":" . $this->current_key ."\n";
	}		

    public function valid() { 
		if($this->last_key !== "\t" && $this->last_key !== $this->current_key){		
			$this->is_newline = true;
			return false;
		}
		return !$this->is_eof;
		//return $this->is_valid;
	} 
}

abstract class BaseReducer{
	public function setup(){}
	public abstract function reduce($key, $values);
	public function cleanup(){}
	public function parse($line){
		return explode("\t", $line, 2);
	}
}

function run(BaseReducer $reducer){
	$values = new ValuesIterator($reducer);
	while(!feof(STDIN)){
		if(!$values->next_key()) break;
		$reducer->reduce($values->key(), $values);
	}

}

